notebooks/utils.py

```diff
@@ -0,0 +1,127 @@
+import torch
+from torch import nn
+from torch import optim
+from matplotlib import pyplot as plt
+import numpy as np
+import seaborn as sns
+
+class Swarm_Simulator:
+    def __init__(self):
+        
+        self.simple_funcs = {
+            'exponential':torch.exp,
+            'sinusoidal':torch.sin
+        }
+        
+        self.debug = False
+        
+        self.params = {
+            'domain_min': -1,
+            'domain_max': 3,
+            'num_points': 81,
+            'function': 'exponential',
+            'hidden': 1,
+            'width': 5,
+            'nepoch': 100,
+            'lr': 0.002,
+            'momentum': 0.9,
+            'swarm_size': 20
+            
+        }
+        
+        self.simulations = []
+        
+        
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        
+        
+    def reset_swarm(self):
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+    def rebuild_predictor(self):
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        self.reset_swarm()
+        
+
+    def make_net(self, hidden_depth, width):
```

Comment:
As this doesn't actually need access to the object I prefer leaving these as functions.

Reply:
I need to have a general better understanding of what goes in and out of a class then. Will mull on this some more.